Improve FlexDLL build tools selection with FlexDLL 0.44 build changes

MisterDA opened the pull request:
I want to use these for a bigger project but I think this patch series can be reviewed on its own. Patches (2) and (3) stem from https://github.com/ocaml/flexdll/pull/153, a PR that's been released as of FlexDLL 0.44.

1. Mark `FLEXLINKFLAGS` as precious and pass them to flexlink invocation.
   Flexlink reads the `FLEXLINKFLAGS` env var for supplementary flags. They could be used when building the compiler. If we don't add them explicitly to `$flexlinkflags` they're:

   1. not shown in the flexlink invocations;
   2. not retained after the build, when the installed ocaml calls flexlink as a linker.

   Prefer passing the `FLEXLINKFLAGS` explicitly in the invocation, at the risk of duplicating the flags if `FLEXLINKFLAGS` is set in the environment rather than given to configure as an argument.
   This change also marks the variable as precious so that any change to it invalidates the build.

2. Define RC to choose an alternative resource compiler.
   This allow overriding the default resource processor, used during FlexDLL's bootstrap. The variable name is defined by libtool.
   Pass the variable to the flexlink Makefile.

3. Pass `$CC` in the flexdll build env.
   FlexDLL's Makefile has its own mechanism to discover the C compiler. Override it with the selected C compiler.

cc @dra27 @nojb @shym

dra27 committed:
Use clang-cl for flexdll support objects

MisterDA committed:
Define MANIFEST_TOOL to choose an alternative manifest generator

This allows using `llvm-mt` instead of `mt.exe`. The MANIFEST_TOOL
variable is defined by libtool.

MisterDA committed:
Mark FLEXLINKFLAGS as precious and pass them to flexlink invocation

Flexlink reads the FLEXLINKFLAGS env var for supplementary flags. They
could be used when building the compiler. If we don't add them
explicitly to `$flexlinkflags` they're:

1. not shown in the flexlink invocations;
2. not retained after the build, when the installed ocaml calls
   flexlink as a linker.

Prefer passing the FLEXLINKFLAGS explicitly in the invocation, at the
risk of duplicating the flags if FLEXLINKFLAGS is set in the
environment rather than given to configure as an argument.

This change also marks the variable as precious so that any change to
it invalidates the build.

MisterDA committed:
Pass $CC in the flexdll build env

FlexDLL's Makefile has its own mechanism to discover the C
compiler. Override it with the selected C compiler.

MisterDA committed:
Define RC to choose an alternative resource compiler

This allow overriding the default resource processor, used during
FlexDLL's bootstrap. The variable name is defined by libtool.

Pass the variable to the flexlink Makefile.

MisterDA committed:
.gitmodules: add the branch config entry

It makes it clearer how to select a different branch than master when
experimenting.
The file doesn't need a copyright header.

dra27 commented:
Full review to come, but while I spot it - please update `ocaml-variants.opam` with a new lowerbound for `"flexdll"` as the build now depends on newer features.

MisterDA committed:
ocaml-variants.opam: require FlexDLL >= 0.44

FlexDLL before 0.44 would try to rediscover the C compiler and other
tools it should use. This would introduce a mismatch if users chose an
C compiler when building OCaml, and FlexDLL would chose another
compiler.

As of FlexDLL 0.44, its Makefile exposes all of the required
settings. OCaml build scripts now know how to forward the selected
tools to FlexDLL build scripts. Update the lower bound to set that in
stone.

MisterDA committed:
Update changes

MisterDA commented:
We could apply the `run-crosscompiler-tests` label to this PR.

dra27 commented on Makefile:
```diff
@@ -634,6 +634,20 @@ FLEXLINK_BUILD_ENV = \
+FLEXLINK_BUILD_ENV += RC="$(RC)"
+endif
+ifeq ($(FLEXDLL_CHAIN),cygwin64)
+FLEXLINK_BUILD_END += CYG64CC="$(CC)"
```
```suggestion
FLEXLINK_BUILD_ENV += CYG64CC="$(CC)"
```

dra27 reviewed (commented):
What's the use-case for `FLEXLINKFLAGS`? I'm not sure about it as a mechanism (thinking with reference to the use `CPPFLAGS`, `CFLAGS` and `LDFLAGS` in terms of hardening) - I'd be more tempted for us to be ensuring `FLEXLINKFLAGS` in the build is unset?

I have a small patch in https://github.com/dra27/ocaml/pull/233/commits/cdbf54419f942df9f2bf32f4bef0b1d35fe9608d which includes the fix to `testsuite/tools/testRelocation.ml` implied by this PR. Also, I'm not sure it's worth worrying about only adding the `MSVCC=`, `MSVC64=` etc. strictly if it's the correct port - why not just unconditionally add all 5 (as I did for the two MSVC ones in that commit)?

nojb commented:
> What's the use-case for FLEXLINKFLAGS?

FWIW, I know of a commecial user of OCaml on Windows which uses `FLEXLINKFLAGS` to link a resource file (`.res`) to all generated artifacts, by setting `FLEXLINKFLAG="-link $resfile"`.

dra27 commented:
Ah, I worded it poorly - I meant using it this way while building OCaml itself, rather than whether the feature in flexlink was used (or is that also what you mean??)

nojb commented:
> Ah, I worded it poorly - I meant using it this way while building OCaml itself, rather than whether the feature in flexlink was used (or is that also what you mean??)

Maybe we are saying the same thing, but just to be clear: the use case I mentioned is intended for linking those `.res` files in the compiler binaries themselves...

MisterDA commented:
THank you both for your input!

> What's the use-case for `FLEXLINKFLAGS`?

I have experimental patches to swap the linker: I can replace the default `ld` and `link` with `ld.lld` or `lld-link`, two linkers provided by LLVM. I use `FLEXLINKFLAGS` to pass the `-use-linker` and `-use-mt` options of FlexDLL 0.44.

> I'm not sure about it as a mechanism (thinking with reference to the use `CPPFLAGS`, `CFLAGS` and `LDFLAGS` in terms of hardening) - I'd be more tempted for us to be ensuring `FLEXLINKFLAGS` in the build is unset?

That's a good point that I can explore!
I don't think that `FLEXLINKFLAGS` need to be unset, as nojb noted, there are existing use cases. But as you've noted, it's likely that I don't need it for for my project. I think it's still a good idea to document it and set it precious, so that a change in the environment invalidates the build.

> I have a small patch in [dra27@cdbf544](https://github.com/dra27/ocaml/commit/cdbf54419f942df9f2bf32f4bef0b1d35fe9608d) which includes the fix to `testsuite/tools/testRelocation.ml` implied by this PR. Also, I'm not sure it's worth worrying about only adding the `MSVCC=`, `MSVC64=` etc. strictly if it's the correct port - why not just unconditionally add all 5 (as I did for the two MSVC ones in that commit)?

Thanks for the patch! If there's really no downside to unconditionally add all 5 variables, that's fine by me, otherwise I'm in favour of avoiding to add too much and rather keep the strict minimum in the environment.

MisterDA committed:
Don't persist FLEXLINKFLAGS in flexlink_flags

dra27 reviewed (approved):
I’m convinced about `FLEXLINKFLAGS`, thanks - this all LGTM

MisterDA commented:
I've rebased the PR. I think it looks good to me too.
